### PR TITLE
add login controller

### PR DESCRIPTION
**Pull request by AbdallahGot**
1. email and password validation.
2. check if the email exists or not.
3. if the email exists check if the password is correct or not.
4. add test.

### CODE REVIEW

File: server/database/models/getDataUsingEmail.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+const { User } = require('./user');
+
+const getEmail = (userEmail) => User.findOne({ where: { email: userEmail } });
```
Comment (LinaYahya) [inline server/database/models/getDataUsingEmail.js:3]: all we're doing here is to call findOne from user, so there is no need to create  a separate function for it, 
In the place you want to find a user call your method there

File: server/controllers/login/index.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+const { compare } = require('bcrypt');
+const { getEmail } = require('../../database/models');
+const signToken = require('../../utils/signToken');
+const { loginValidation } = require('../../utils/validation');
+const { customError } = require('../errors');
+
+const loginHandler = async (req, res, next) => {
+  try {
+    const validateValue = await loginValidation(req.body);
+    const checkedEmail = await getEmail(validateValue.email);
```
Comment (LinaYahya) [inline server/controllers/login/index.js:10]:
```suggestion
    const {id, password: hashedPassword} = await User.findOne({ where: { email: validateValue.email } });
```

File: server/controllers/login/index.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+const { compare } = require('bcrypt');
+const { getEmail } = require('../../database/models');
+const signToken = require('../../utils/signToken');
+const { loginValidation } = require('../../utils/validation');
+const { customError } = require('../errors');
+
+const loginHandler = async (req, res, next) => {
+  try {
+    const validateValue = await loginValidation(req.body);
```
Comment (LinaYahya) [inline server/controllers/login/index.js:9]: destructure email and password from there
```suggestion
    const {email, password} = await loginValidation(req.body);
```

File: server/utils/signToken.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+const { sign } = require('jsonwebtoken');
+
+module.exports = (userId) => new Promise((resolve, reject) => {
```
Comment (LinaYahya) [inline server/utils/signToken.js:3]: You can put it in the  `jwt` file

File: server/controllers/login/index.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    const checkedEmail = await getEmail(validateValue.email);
+    if (!checkedEmail) throw customError('incorrect password or email ...', 401);
+    const checkedPassword = await compare(validateValue.password, checkedEmail.password);
+    if (!checkedPassword) throw customError('incorrect password or email ...', 401);
+    const signTokenCookie = await signToken(checkedEmail.id);
+    if (signTokenCookie) res.cookie('userToken', signTokenCookie);
+    res.json({ msg: 'logged in successfully ..', data: { id: checkedEmail.id } });
+  } catch (err) {
+    if (err.name === 'ValidationError') next(customError(err.message, 401));
```
Comment (LinaYahya) [inline server/controllers/login/index.js:18]:
```suggestion
    if (err.name === 'ValidationError') next(customError(err.message, 400));
```

File: server/controllers/login/index.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+const loginHandler = async (req, res, next) => {
+  try {
+    const validateValue = await loginValidation(req.body);
+    const checkedEmail = await getEmail(validateValue.email);
+    if (!checkedEmail) throw customError('incorrect password or email ...', 401);
+    const checkedPassword = await compare(validateValue.password, checkedEmail.password);
+    if (!checkedPassword) throw customError('incorrect password or email ...', 401);
+    const signTokenCookie = await signToken(checkedEmail.id);
+    if (signTokenCookie) res.cookie('userToken', signTokenCookie);
```
Comment (LinaYahya) [inline server/controllers/login/index.js:15]: there's always a signToken, So need to check if exist or not,

File: server/controllers/login/index.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    res.json({ msg: 'logged in successfully ..', data: { id: checkedEmail.id } });
```
Comment (LinaYahya) [inline server/controllers/login/index.js:16]: also send user name here, which is a firstName + ' ' + lastName

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

File: server/controllers/login/index.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+const { compare } = require('bcrypt');
+const { User } = require('../../database/models');
+const { signToken } = require('../../utils');
+const { loginValidation } = require('../../utils/validation');
+const { customError } = require('../errors');
+
+const loginHandler = async (req, res, next) => {
+  try {
+    const { email: userEmail, password: hashedPassword } = await loginValidation(req.body);
```
Comment (LinaYahya) [inline server/controllers/login/index.js:11]:
```suggestion
    const { email: userEmail, password } = await loginValidation(req.body);
```

File: server/controllers/login/index.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    const userData = await User.findOne({
+      where: { email: userEmail },
+    });
+    if (!userData) {
+      throw customError('incorrect password or email ...', 401);
+    }
+    const {
+      id, password, first_name, last_name,
```
Comment (LinaYahya) [inline server/controllers/login/index.js:19]:
```suggestion
      id, password: hashedPassword, first_name, last_name,
```

File: server/controllers/login/index.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    } = userData;
+    const providerName = `${first_name} ${last_name}`;
+    const checkedPassword = await compare(hashedPassword, password);
```
Comment (LinaYahya) [inline server/controllers/login/index.js:22]:
```suggestion
    const checkedPassword = await compare(password, hashedPassword);
```

File: server/utils/jwt.js [hunk 11]
```diff
@@ -1,10 +1,18 @@
...
+const { verify, sign } = require('jsonwebtoken');
 
 const verifyToken = (token) => new Promise((resolve, reject) => {
   verify(token, process.env.SECRET_KEY, (err, match) => {
     if (err) return reject(err);
     return resolve(match);
   });
 });
+const signToken = (providerName, userId) => new Promise((resolve, reject) => {
+  sign({ providerName, providerID: userId, role: 'provider' }, process.env.SECRET_KEY, (err, token) => {
```
Comment (LinaYahya) [inline server/utils/jwt.js:10]:
```suggestion
  sign({ providerName, providerID: userId}, process.env.SECRET_KEY, (err, token) => {
```
No more need for the role here

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by RandSohail — APPROVED** [review]

**Review by devstm — APPROVED** [review]

**Review by abdulrahman-2020 — APPROVED** [review]

**Review by LinaYahya — APPROVED** [review]

**Pull request merged**